---
title: "Input data"
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

### Data with gene ID information

<ul>
<li>text files expected</li>
<li> see below for the expected format</li>
</ul>

##### Gene position information 

```{r gff}
# from main_settings.R
entrezDT_file <- paste0("gff_entrez_position_GRCh37p13_nodup.txt")
gff_dt <- read.delim(entrezDT_file, header = TRUE, stringsAsFactors = FALSE)
head(gff_dt)
```

##### Older entrez IDs mapping

```{r history}
# from main_settings.R
historyDT_file <- file.path("gene_history_reordered.txt")
historyDT <- read.delim(historyDT_file,
                          stringsAsFactors = FALSE,
                          header=TRUE)
head(historyDT)
```

##### Gene symbol - entrez ID mapping

```{r symbol}
# from main_settings.R
symbolDT_file <- paste0("final_entrez2syno.txt")
symbolDT <- read.delim(symbolDT_file,
                          stringsAsFactors = FALSE,
                          header=TRUE)
head(symbolDT)
```

##### Gene ensemble ID - entrez ID mapping

```{r ensembl}
# from main_settings.R
ensemblDT_file <- paste0("final_entrez2ensembl.txt")
ensemblDT <- read.delim(ensemblDT_file,
                          stringsAsFactors = FALSE,
                          header=TRUE)
head(ensemblDT)
```

### TAD positions and gene-to-TAD assignment

<ul>
<li>text files expected</li>
<li>4-column TAD positions: chromosome/region/start/end</li>
<li>5-column gene information: geneID/chromosome/start/end/region</li>
<li>region label is expected to be like <em>chr<Nbr>_TAD<nbr></em></li>
</ul>

(order of the columns and extra-columns do not matter)

```{r TAD_and_genes}
# TAD positions
TADpos_file <- file.path("..", "EXAMPLE", "DATA", "ENCSR489OCU_NCI-H460_all_assigned_regions.txt")
tad_pos_dt <- read.delim(TADpos_file,
                          stringsAsFactors = FALSE,
                          header=FALSE, 
                          col.names=c("chromosome", "region", "start", "end"))
head(tad_pos_dt)

# Gene positions and gene-to-TAD assignment
gene2tadDT_file <- file.path("..", "EXAMPLE", "DATA", "ENCSR489OCU_NCI-H460_all_genes_positions.txt")
gene2tad_dt <- read.delim(gene2tadDT_file,
                          stringsAsFactors = FALSE,
                          header=FALSE, 
                          col.names=c("entrezID", "chromosome", "start", "end", "region"))
gene2tad_dt$entrezID <- as.character(gene2tad_dt$entrezID)
head(gene2tad_dt)
```


### Gene expression data

<ul>
<li> RData format expected (i.e. a format that can be imported with <em>load()</em>)</li>
<li>row names are expected to be gene IDs</li>
<li>column names are expected to be sample IDs</li>
<li>can contain extra columns (only those corresponding to sample1 and sample2 IDs will be retained)</li>
</ul>

```{r geneexpr}
# used to create classes of gene expression (step 5fc)
rna_fpkmDT_file <- file.path("..", "EXAMPLE", "DATA", "TCGAluad_norm_luad_fpkmDT.RData")
rna_fpkmDT <- get(load(rna_fpkmDT_file))
rna_fpkmDT[1:5,1:5]

# used for gene-level differential expression analysis (step)
rnaseqDT_file <- file.path("..", "EXAMPLE", "DATA", "TCGAluad_norm_luad_rnaseqDT_v2.RData")
rnaseqDT <- get(load(rnaseqDT_file))
rnaseqDT[1:5,1:5]

```


### Sample information

<ul>
<li>RData format expected (i.e. a format that can be imported with <em>load()</em>)</li>
<li>should be contained in the columns of the gene expression data</li>
<li><em>sample1_file</em> should store sample IDs for <em>cond1</em> and <em>sample2_file</em> those of <em>cond2</em></li>
</ul>

```{r sample}
# samples that correspond to cond1 <- <CONDITION_1>
sample1_file <- file.path("..", "EXAMPLE", "DATA", "norm_ID.RData")
samp1 <- get(load(sample1_file))
head(samp1)
# samples that correspond to cond2 <- <CONDITION_2>
sample2_file <- file.path("..", "EXAMPLE", "DATA", "luad_ID.RData")
samp2 <- get(load(sample2_file))
head(samp2)
```



### Import correlation permutation values 

For step 9, case of a provided file:

<ul>
<li>RData format expected (i.e. a format that can be imported with <em>load()</em>)</li>
<li>this is a list of list
<li>each element of the outer list should correspond to permutation value for a given TAD</li>
<li>in the inner list, there should be one element named <em>meanCorr</em></li>
</ul>

```{r correlation}
all_permutCorr_data <- get(load(file.path("all_sample_corrValues.RData")))
str(all_permutCorr_data[[1]])
```







